### PR TITLE
Move table restructureing to R-SQL lesson using tidyr

### PR DESCRIPTION
**Pull request by brymz**
As per #213

### CODE REVIEW

**Comment by brymz** [comment]
Ready for review.

Code solution: https://github.com/weecology/datacarp-semester-biology-exercise-solns/pull/107

File: exercises/Tidyr-tree-biomass-R.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+We're going to estimate the total tree biomass for trees in a 96 hectare area of 
+the Western Ghats in India. [The raw data](http://esapubs.org/archive/ecol/E091/216/Macroplot_data_Rev.txt) is available on [Ecological Archives](http://esapubs.org/Archive/). Unfortunately, the data is stored in poor database 
+structure and using all of the tree stems would be difficult without first 
+tidying up the data. You can have a look at [the metadata](http://esapubs.org/archive/ecol/E091/216/metadata.htm) to get familiar with the data structure.
+
+1. Use `tidyr` to `gather()` the raw data into rows for each measured stem. 
+2. Write a function that takes a vector of tree diameters as an argument and   
+returns a vector of tree masses.
+3. Stems are measured in girth (*or circumference*) rather than diameter. Write a function to take an vector of circumferences as an argument and returns a vector of diameters (*circumference = pi \* diameter*).
```
Comment (ethanwhite) [inline exercises/Tidyr-tree-biomass-R.md]: "that takes a vector"

File: exercises/Tidyr-tree-biomass-R.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+topic: Tidyr
+title: Tree Biomass
+language: R
+---
+
+Estimating the total amount of biomass (*the total mass of all individuals*) 
+in forests is important for understanding the global carbon budget and how the 
+earth will respond to increases in carbon dioxide emissions. Measuring the mass 
+of whole trees is a major effort and requires destructive harvest of the tree. 
+It's pretty much impossible to weigh an entire forest (*even if we were willing
```
Comment (ethanwhite) [inline exercises/Tidyr-tree-biomass-R.md]: Let's drop this sentence and the parenthetical. This stuff can be fun, but I think it makes the material more difficult for students whose first language isn't English.

File: exercises/Tidyr-tree-biomass-R.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+There are lots of equations for estimating the mass of a tree from its diameter, 
+but one good option is the equation: 
+    
+Mass = 0.124 * Diameter<sup>2.53</sup>
+
+where `Mass` is measured in kg of dry above-ground biomass and `Diameter` is in cm [DBH](https://en.wikipedia.org/wiki/Diameter_at_breast_height) ([Brown 1997](http://www.fao.org/docrep/W4095E/W4095E00.htm)). 
+
+We're going to estimate the total tree biomass for trees in a 96 hectare area of 
+the Western Ghats in India. [The raw data](http://esapubs.org/archive/ecol/E091/216/Macroplot_data_Rev.txt) is available on [Ecological Archives](http://esapubs.org/Archive/). Unfortunately, the data is stored in poor database
```
Comment (ethanwhite) [inline exercises/Tidyr-tree-biomass-R.md]: "stored using untidy database structure"

Comment (ethanwhite) [inline exercises/Tidyr-tree-biomass-R.md]: Just tweaked that last suggestion to just clean up your original version.

File: exercises/Tidyr-tree-biomass-R.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+2. Write a function that takes a vector of tree diameters as an argument and   
+returns a vector of tree masses.
+3. Stems are measured in girth (*or circumference*) rather than diameter. Write a function to take an vector of circumferences as an argument and returns a vector of diameters (*circumference = pi \* diameter*).
+4. Use the two functions you've written to estimate the total biomass (*i.e., 
+the sum of the masses*) of trees in this dataset and print the result to the
+screen.
+5. `separate()` the `SpCode` into `GenusCode` and `SpEpCode` and estimate the 
+total biomass per genus in a table. *It is a bit of an erroneous assumption that 
+all of the genus in the dataset will be identified uniquely by their four letter 
+code, but we'll go with it for the purpose of this exercise.*
```
Comment (ethanwhite) [inline exercises/Tidyr-tree-biomass-R.md]: "*Technically the four letter code doesn't uniquely identify all of the genera in the datasets, but we'll assume it does for the purpose of this exercise*"

File: materials/tidyr.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+1. Order doesn’t matter 
+2. No duplicate rows
+3. Every cell contains one value
+4. One column per type of information
+5. No redundant information
+
+### Restructure tables with messy data 
+
+* Cells with multiple values breaks rule #3.
```
Comment (ethanwhite) [inline materials/tidyr.md]: breaks -> break

File: materials/tidyr.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+```
+> scary_sightings
+  animals brick_road emerald_city
+1   lions        1-Y         17-N
+2  tigers        0-N          8-Y
+3   bears        0-N         64-N
+```
+
```
Comment (ethanwhite) [inline materials/tidyr.md]: Add a brief explanation of what the values in the tables represent.

File: materials/tidyr.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        * `Y` or `N`
+
+### `tidyr` helps restructure messy data
+
+* `gather()`
+    * Removes redundant columns
+    * Arguments:
+        * Piped `data.frame`
+        * Column names
+        * Column range
```
Comment (ethanwhite) [inline materials/tidyr.md]: I think we should be a little clearer here about what the column names and range are and separate the column name for the column that holds the headers from the one that holds the values.

**Comment by ethanwhite** [comment]
Looks great, just some minor cleanup.

File: solutions/Scientific-tree-biomass-challenge-R.txt [hunk 8]
```diff
@@ -1 +1 @@
-[1] 151992268
\ No newline at end of file
+[1] 26710042
```
Comment (ethanwhite) [inline solutions/Scientific-tree-biomass-challenge-R.txt]: There should be output for the full forest and for each genera.

File: materials/tidyr.md (hunk 6 above)
Comment (brymz) [inline materials/tidyr.md]: explanation added

File: solutions/Scientific-tree-biomass-challenge-R.txt (hunk 8 above)
Comment (brymz) [inline solutions/Scientific-tree-biomass-challenge-R.txt]: This is the output solution for http://www.datacarpentry.org/semester-biology/exercises/Scientific-tree-biomass-challenge-R/. I found the solution to be wrong during the revision of this exercise to `Tidyr-tree-biomass-R.R`. This exercise only asks for the single value of total forest biomass.

**Comment by brymz** [comment]
Revisions complete. Ready to go.

**Pull request merged**